AX: Add the ability to dump the accessibility tree via notifyutil

twilco opened the pull request:
#### 3bcec6c1218af55ab8f6f6eeee13cc9f12e6bdc9
<pre>
AX: Add the ability to dump the accessibility tree via notifyutil
<a href="https://bugs.webkit.org/show_bug.cgi?id=295101">https://bugs.webkit.org/show_bug.cgi?id=295101</a>
<a href="https://rdar.apple.com/154490712">rdar://154490712</a>

Reviewed by Joshua Hoffman.

After this commit, with a build of WebKit that has ENABLE(TREE_DEBUGGING), one can run:

notifyutil -p com.apple.WebKit.showAccessibilityTree

To dump the accessibility tree on-demand.

Also, when ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE), this commit changes streamAXCoreObject to print the parent object
(including ignored), and streamSubtree to print ignored children.

* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::parentInCoreTree const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::streamSubtree):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::showAccessibilityTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::parentInCoreTree const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNode):
* Source/WebCore/page/cocoa/PageCocoa.mm:
(WebCore::Page::platformInitialize):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::printAccessibilityTreeForLiveDocuments):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebKit/Resources/cocoa/NotificationAllowList/ForwardedNotifications.def:

Canonical link: <a href="https://commits.webkit.org/296746@main">https://commits.webkit.org/296746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dca042290f763006688325f92656eb5f4bac1d21

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109448 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19535 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114653 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59677 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111411 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37694 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83187 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112396 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98582 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63646 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23106 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16725 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59272 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93094 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16767 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117767 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36489 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27015 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92198 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36860 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94844 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/92035 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23431 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36948 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14693 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32293 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36382 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41858 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36053 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39392 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37757 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -1053,4 +1053,6 @@ bool isVisibilityHidden(const RenderStyle&);
 
 WTF::TextStream& operator<<(WTF::TextStream&, AXNotification);
 
+void showAccessibilityTree(Document&);
```
How about dumpAccessibilityTreeToStderr or something more specific? Show makes it sound like it will be visual

twilco replied:
Yeah agreed. The reason I chose this name is to follow the pattern of the other `notifyutil` tree dump functions, e.g.:

[showLayerTree](https://github.com/WebKit/WebKit/blob/dd9bf16cd3995f89a44871d8338927691b0ee68c/Source/WebCore/rendering/RenderObject.cpp#L3032)

[showRenderTree](https://github.com/WebKit/WebKit/blob/dd9bf16cd3995f89a44871d8338927691b0ee68c/Source/WebCore/rendering/RenderObject.cpp#L3020)

[showPaintOrderTree](https://github.com/WebKit/WebKit/blob/dd9bf16cd3995f89a44871d8338927691b0ee68c/Source/WebCore/rendering/RenderObject.cpp#L3008)

But I don't think we really _need_ to follow the pattern in this case, so I'll change it.

minorninth commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -3033,6 +3033,21 @@ void printLayerTreeForLiveDocuments()
+            continue;
+        if (document->frame()) {
+            if (document->frame()->isRootFrame())
+                WTFLogAlways("Accessibility tree for root document %p %s", document.ptr(), document->url().string().utf8().data());
```
Does WTFLogAlways do the same as fprintf to stderr, with the same buffering and everything? I think it'd be best if all of the code that logs uses the same mechanism

twilco replied:
WTFLogAlways is a thin-wrapper over this:

```cpp
// Assertions.cpp
void WTFLogAlwaysV(const char* format, va_list args)
{
    vprintf_stderr_with_trailing_newline(nullptr, format, args);
}
```

So seems like it's always stderr?

I mostly chose `WTFLogAlways` to follow the pattern of the several other similar "dump tree" methods, e.g.:

[printLayerTreeForLiveDocuments](https://github.com/WebKit/WebKit/blob/dd9bf16cd3995f89a44871d8338927691b0ee68c/Source/WebCore/rendering/RenderObject.cpp#L3031)

[printRenderTreeForLiveDocuments](https://github.com/WebKit/WebKit/blob/dd9bf16cd3995f89a44871d8338927691b0ee68c/Source/WebCore/rendering/RenderObject.cpp#L3019)

[printPaintOrderTreeForLiveDocuments](https://github.com/WebKit/WebKit/blob/dd9bf16cd3995f89a44871d8338927691b0ee68c/Source/WebCore/rendering/RenderObject.cpp#L3007)

...and a few others

Would you prefer we explicitly `fprintf(stderr, "");` here?

minorninth replied:
If they're both directly printing to stderr then it's fine. I've just been bitten by, e.g. mixing cerr and stderr because they have separate buffering.

minorninth reviewed (commented):
Love it!

twilco committed:
AX: Add the ability to dump the accessibility tree via notifyutil
https://bugs.webkit.org/show_bug.cgi?id=295101
rdar://154490712

Reviewed by Joshua Hoffman.

After this commit, with a build of WebKit that has ENABLE(TREE_DEBUGGING), one can run:

notifyutil -p com.apple.WebKit.showAccessibilityTree

To dump the accessibility tree on-demand.

Also, when ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE), this commit changes streamAXCoreObject to print the parent object
(including ignored), and streamSubtree to print ignored children.

* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::parentInCoreTree const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::streamSubtree):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::showAccessibilityTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::parentInCoreTree const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNode):
* Source/WebCore/page/cocoa/PageCocoa.mm:
(WebCore::Page::platformInitialize):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::printAccessibilityTreeForLiveDocuments):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebKit/Resources/cocoa/NotificationAllowList/ForwardedNotifications.def:

Canonical link: https://commits.webkit.org/296746@main
